File: vscode/src/classify/prompt.ts
/**
 * Prompt building for Claude classification.
 * Ported from Python human_review/cli.py:_build_classify_prompt()
 */

import type { ChangedFile, DiffHunk } from "../state/types";
import { TRUST_PATTERNS } from "../trust/patterns";
import { getHunkKey } from "../utils/hash";

/**
 * Build the diff content for classification.
 */
export function buildDiffContent(
  hunksToClassify: Array<{ file: ChangedFile; hunk: DiffHunk; hunkKey: string }>,
): string {
  const lines: string[] = [];

  for (const { file, hunk, hunkKey } of hunksToClassify) {
    lines.push(`=== ${hunkKey} ===`);
    lines.push(`File: ${file.path} (${file.status})`);
    lines.push(hunk.header);
    lines.push(hunk.content);
    lines.push("");
  }

  return lines.join("\n");
}

/**
 * Build the full classification prompt for Claude.
 */
export function buildClassifyPrompt(diffContent: string): string {
  // Build pattern list
  const patternList = Object.values(TRUST_PATTERNS)
    .map((p) => `- \`${p.id}\` — ${p.description}`)
    .join("\n");

  return `Classify each hunk in this diff. For each hunk, provide:
1. **label**: Array of trust patterns from the taxonomy (can be empty if no patterns apply)
2. **reasoning**: Brief explanation of what the change does

## Trust Patterns Taxonomy

Only use patterns from this list. Leave label empty if no patterns apply.

${patternList}

## Rules

- Apply patterns ONLY when they FULLY describe the change
- If a hunk has mixed changes (e.g., imports + logic), leave label empty
- Multiple patterns are allowed if the hunk combines trustable changes
- Reasoning should be specific and clear (e.g., "Added import for ChoicesFieldMixin")

## Output Format

Return a JSON object mapping hunk_key to classification:

\`\`\`json
{
  "filepath:hash": {
    "label": ["pattern:id"],
    "reasoning": "Brief explanation"
  }
}
\`\`\`

## Diff to Classify

${diffContent}

Return ONLY the JSON object, no other text.`;
}

/**
 * Get unlabeled hunks from files that need classification.
 */
export function getUnlabeledHunks(
  files: ChangedFile[],
  hunks: Record<string, { reasoning: string | null }>,
): Array<{ file: ChangedFile; hunk: DiffHunk; hunkKey: string }> {
  const unlabeled: Array<{ file: ChangedFile; hunk: DiffHunk; hunkKey: string }> = [];

  for (const file of files) {
    for (const hunk of file.hunks) {
      const hunkKey = getHunkKey(hunk.filePath, hunk.hash);
      const hunkState = hunks[hunkKey];
      if (!hunkState || hunkState.reasoning === null) {
        unlabeled.push({ file, hunk, hunkKey });
      }
    }
  }

  return unlabeled;
}

File: review/src/classify/prompt.rs
use crate::trust::patterns::get_trust_taxonomy;
use std::fmt::Write;

/// Labels that are fully handled by the static classifier and should
/// never be offered to the AI (to prevent hallucinated matches).
const STATIC_ONLY_LABELS: &[&str] = &["formatting:whitespace", "generated:lockfile"];

/// Build a flat list of all valid labels with descriptions,
/// excluding labels that are handled entirely by the static classifier.
fn build_taxonomy_string() -> String {
    let taxonomy = get_trust_taxonomy();
    let mut result = String::new();

    for category in taxonomy {
        for pattern in category.patterns {
            if STATIC_ONLY_LABELS.contains(&pattern.id.as_str()) {
                continue;
            }
            let _ = writeln!(result, "- `{}`: {}", pattern.id, pattern.description);
        }
    }

    result
}

/// Input structure for a hunk to classify
#[derive(Debug, Clone, serde::Deserialize)]
pub struct HunkInput {
    pub id: String,
    #[serde(rename = "filePath")]
    pub file_path: String,
    pub content: String,
}

/// Build a prompt for classifying a single hunk
pub fn build_single_hunk_prompt(hunk: &HunkInput) -> String {
    let taxonomy = build_taxonomy_string();

    format!(
        r#"Determine if this diff hunk is a trivial, mechanical change that a reviewer can safely skip. If it matches a trivial pattern, apply the matching label. If not, return an empty label array.

# Valid Labels (use ONLY these exact strings)

{taxonomy}
# Rules

1. DEFAULT TO EMPTY LABELS. Most hunks need human review. Only apply a label for trivial, mechanical changes.
2. A label applies ONLY when the ENTIRE hunk matches its description exactly.
3. Any change to values, logic, behavior, or configuration = empty labels.
4. Mixed changes (e.g., import added + code changed) = empty labels.
5. If a hunk changes code AND adds/modifies/removes a comment, it is mixed = empty labels.
6. Use ONLY the exact label strings listed above.

# Hunk

File: {file_path}
```diff
{content}
```

# Response

STEP 1: List each changed line (+ or -) and what it does (code, comment, whitespace, import, etc.)
STEP 2: Do ALL changed lines fall under a single trivial label's description?
STEP 3: If yes, use that label. If not, return empty labels.

After your analysis, return JSON on its own line:
{{"label": [], "reasoning": "one sentence"}}"#,
        taxonomy = taxonomy,
        file_path = hunk.file_path,
        content = hunk.content
    )
}

/// Build a prompt for classifying multiple hunks in a single request
pub fn build_batch_prompt(hunks: &[HunkInput]) -> String {
    let taxonomy = build_taxonomy_string();

    let mut hunks_section = String::new();
    for (i, hunk) in hunks.iter().enumerate() {
        let _ = write!(
            hunks_section,
            r"### Hunk {} (ID: {})
File: {}
```diff
{}
```

",
            i + 1,
            hunk.id,
            hunk.file_path,
            hunk.content
        );
    }

    format!(
        r#"Determine if these diff hunks are trivial, mechanical changes that a reviewer can safely skip. If a hunk matches a trivial pattern, apply the matching label. If not, return an empty label array.

# Valid Labels (use ONLY these exact strings)

{taxonomy}
# Rules

1. DEFAULT TO EMPTY LABELS. Most hunks need human review. Only apply a label for trivial, mechanical changes.
2. A label applies ONLY when the ENTIRE hunk matches its description exactly.
3. Any change to values, logic, behavior, or configuration = empty labels.
4. Mixed changes (e.g., import added + code changed) = empty labels.
5. If a hunk changes code AND adds/modifies/removes a comment, it is mixed = empty labels.
6. Use ONLY the exact label strings listed above.
7. You MUST classify EVERY hunk ID listed above.

# Hunks

{hunks_section}
# Response

For EACH hunk, analyze it step by step:
STEP 1: List each changed line (+ or -) and what it does (code, comment, whitespace, import, etc.)
STEP 2: Do ALL changed lines fall under a single trivial label's description?
STEP 3: If yes, use that label. If not, return empty labels.

After analyzing all hunks, return JSON on its own line:
{{
  "hunk_id": {{"label": [], "reasoning": "one sentence"}},
  ...
}}"#
    )
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_taxonomy_string_excludes_static_only_labels() {
        let taxonomy = build_taxonomy_string();
        assert!(
            !taxonomy.contains("formatting:whitespace"),
            "formatting:whitespace should be excluded from AI taxonomy"
        );
        assert!(
            !taxonomy.contains("generated:lockfile"),
            "generated:lockfile should be excluded from AI taxonomy"
        );
        // Other labels should still be present
        assert!(taxonomy.contains("formatting:line-length"));
        assert!(taxonomy.contains("formatting:style"));
        assert!(taxonomy.contains("imports:added"));
    }
}
